doc/neps/nep-0054-simd-cpp-highway.rst

```diff
@@ -0,0 +1,368 @@
+times for different architectures and picking the one with unsupported
+instructions. To ensure to avoid the issue, it's advisable to keep numerical
+kernels inside the source code and refrain from defining non-inlined functions
+within cache-able objects.
+
+
+C++ refactoring considerations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We want to move from C to C++, which will naturally involve a significant
+amount of refactoring, for two main reasons:
+
+- get rid of the NumPy-specific templating language for more expressive C++
+- this would make using sizeless intrinsics (like for SVE) easier.
+
+In addition, we see the following considerations:
+
+- If we use Highway, we would need to switch the C++ wrappers from universal
+  intrinsics to Highway. On the other hand, the work to move to C++ is not
+  complete.
+- If we use Highway, we'd need to rewrite existing kernels using Highway
+  intrinsics. But again, moving to C++ requires touching all those kernels
+  anyway.
+- One concern regarding Highway was whether it is possible to obtain a function
+  pointer for an architecture-specific function instead of calling that
+  function directly. This so that we can be sure that calling 1-D inner loop
+  many times for a single Python API invocation does not incur the dispatching
+  overhead many times. This was investigated: this can be done with Highway
+  too.
+- A second concern was whether it's possible with Highway to allow the user at
+  runtime to select or disable dispatching to certain instruction sets. This is
+  possible.
+- Use of tags in Highway's C++ implementation reduces code duplication but the
+  added templating makes C-level testing and tracing more complicated.
+
+
+The ``_simd`` unit testing module
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Rewriting the ``_simd testing`` module to use C++ was done very recently in PR
+`gh-24069`_. It depends on the main PR for the move to C++, `gh-21057`_.
+It allows one to access the C++ intrinsics with almost the same signature, but
+from Python. This is a great way not only for testing, but also for designing
+new SIMD kernels.
+
+It may be possible to add a similar testing and prototyping feature to Highway
+(which uses plain ``googletest``), however currently the NumPy way is quite a
+bit nicer.
+
+
+Math routines
+~~~~~~~~~~~~~
+
+Math or numerical routines are written at a higher level of abstraction than
+the universal intrinsics that are the main focus of this NEP. Highway has only
+a limited number of math routines, and they are not precise enough for NumPy's
+needs. So either way, NumPy's existing routines (which use universal
+intrinsics) will stay, and if we go the Highway route they'll simply have to
+use Highway primitives internally. We could still use Highway sorting routines.
+If we do accept lower-precision routines (via a user-supplied choice, i.e.
+extending ``errstate`` to allow a precision option), we could use
+Highway-native routines.
```

Comment:
We are also open to adding new higher-precision function versions to Highway, e.g. as part of Sayed's work to port from SVML.